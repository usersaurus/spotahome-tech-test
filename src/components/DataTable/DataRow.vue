<script setup lang="ts">
import { IHouse } from '../../domain/models/House'

interface IDataRow {
  data: IHouse
}

defineProps<IDataRow>()
</script>

<template>
  <tr class="data-row">
    <td class="data-row--item">
      {{ data.title }}
    </td>
    <td class="data-row--item">
      {{ data.address }}
    </td>
    <td class="data-row--item">
      {{ data.city }}
    </td>
    <td class="data-row--item">
      <a :href="data.image" target="_blank"
        ><i class="fa-solid fa-image"></i
      ></a>
    </td>
    <td class="data-row--item">
      <a :href="data.link" target="_blank">
        <i class="fa-solid fa-up-right-from-square"></i>
      </a>
    </td>
  </tr>
</template>

<style lang="postcss">
.data-row--item {
  padding: 1rem;
  text-transform: capitalize;
}
</style>
