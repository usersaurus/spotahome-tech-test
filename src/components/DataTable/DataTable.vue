<script setup lang="ts">
import { IHouse } from '../../domain/models/House'
import DataRow from './DataRow.vue'

interface IDataTableProps {
  columns: string[]
  data: IHouse[]
  loading: boolean
}

defineProps<IDataTableProps>()
</script>

<template>
  <div v-for="column in columns" :key="column">{{ column }} |</div>
  <DataRow v-for="elemn in data" :key="Object.values(elemn)[0]" :data="elemn" />
</template>

<style></style>
