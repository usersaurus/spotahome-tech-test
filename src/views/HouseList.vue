<script setup lang="ts">
import { onMounted, ref } from 'vue'
import { getHousesList } from '../domain/services'
import { useAppStore } from '../stores/app'
import DataTable from '../components/DataTable/DataTable.vue'

const store = useAppStore()
const columns = ref([] as string[])

onMounted(async () => {
  const housesList = await getHousesList()

  store.setHouseList(housesList)
  columns.value = store.houseList[0] ? Object.keys(store.houseList[0]) : []
})
</script>

<template>
  <div class="house-list">
    <DataTable :columns="columns" :data="store.houseList" :loading="true" />
  </div>
</template>

<style lang="postcss">
.house-list {
  padding: 1rem;
  position: relative;
  max-width: 100%;
}
</style>
