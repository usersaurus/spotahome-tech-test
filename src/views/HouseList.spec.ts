import { mount } from '@vue/test-utils'
import HouseList from './HouseList.vue'
import DataTable from '../components/DataTable/DataTable.vue'
import DataRow from '../components/DataTable/DataRow.vue'
import { setActivePinia, createPinia } from 'pinia'

jest.mock('../api')

describe('The HouseList component', () => {
  setActivePinia(createPinia())

  const wrapper = mount(HouseList)

  it('should render a DataTable component', () => {
    expect(wrapper.findComponent(DataTable).exists()).toBe(true)
  })

  it('should render 2 DataRow components', () => {
    expect(wrapper.findAllComponents(DataRow).length).toBe(2)
  })

  it('should render DataRows with the expected content', () => {
    const data = wrapper.findAllComponents(DataRow)

    expect(data).toMatchSnapshot()
  })
})
