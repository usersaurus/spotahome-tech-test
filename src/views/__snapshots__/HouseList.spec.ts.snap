// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The HouseList component should render DataRows with the expected content 1`] = `
Array [
  VueWrapper {
    "__app": null,
    "__setProps": undefined,
    "componentVM": Object {
      "data": Object {
        "address": "Paseo de la Castellana",
        "city": "madrid",
        "image": "https://d1052pu3rm1xk9.cloudfront.net/fsosw_960_540_verified_ur_6_50/74ef00db848c1ce44ea03677836b5bb61243e3f9d6a9d5a656d3c3cb.jpg",
        "link": "https://www.spotahome.com/madrid/for-rent:apartments/79906?utm_source=housinganywhere&utm_medium=cffeeds&utm_campaign=normalads",
        "title": "madrid - 280.000,00 €",
      },
    },
    "isDisabled": [Function],
    "rootVM": Object {},
    "wrapperElement": <tr
      class="data-row"
    >
      <td
        class="data-row--item"
      >
        madrid - 280.000,00 €
      </td>
      <td
        class="data-row--item"
      >
        Paseo de la Castellana
      </td>
      <td
        class="data-row--item"
      >
        madrid
      </td>
      <td
        class="data-row--item"
      >
        <a
          href="https://d1052pu3rm1xk9.cloudfront.net/fsosw_960_540_verified_ur_6_50/74ef00db848c1ce44ea03677836b5bb61243e3f9d6a9d5a656d3c3cb.jpg"
          target="_blank"
        >
          <i
            class="fa-solid fa-image"
          />
        </a>
      </td>
      <td
        class="data-row--item"
      >
        <a
          href="https://www.spotahome.com/madrid/for-rent:apartments/79906?utm_source=housinganywhere&utm_medium=cffeeds&utm_campaign=normalads"
          target="_blank"
        >
          <i
            class="fa-solid fa-up-right-from-square"
          />
        </a>
      </td>
    </tr>,
  },
  VueWrapper {
    "__app": null,
    "__setProps": undefined,
    "componentVM": Object {
      "data": Object {
        "address": "calle virtudes",
        "city": "madrid",
        "image": "https://d1052pu3rm1xk9.cloudfront.net/fsosw_960_540_verified_ur_6_50/9617b48286d8cf89fe755b19d884901be10f6b613a6fc872be2e83be.jpg",
        "link": "https://www.spotahome.com/madrid/for-rent:apartments/79994?utm_source=housinganywhere&utm_medium=cffeeds&utm_campaign=normalads",
        "title": "madrid - 95.000,00 €",
      },
    },
    "isDisabled": [Function],
    "rootVM": Object {},
    "wrapperElement": <tr
      class="data-row"
    >
      <td
        class="data-row--item"
      >
        madrid - 95.000,00 €
      </td>
      <td
        class="data-row--item"
      >
        calle virtudes
      </td>
      <td
        class="data-row--item"
      >
        madrid
      </td>
      <td
        class="data-row--item"
      >
        <a
          href="https://d1052pu3rm1xk9.cloudfront.net/fsosw_960_540_verified_ur_6_50/9617b48286d8cf89fe755b19d884901be10f6b613a6fc872be2e83be.jpg"
          target="_blank"
        >
          <i
            class="fa-solid fa-image"
          />
        </a>
      </td>
      <td
        class="data-row--item"
      >
        <a
          href="https://www.spotahome.com/madrid/for-rent:apartments/79994?utm_source=housinganywhere&utm_medium=cffeeds&utm_campaign=normalads"
          target="_blank"
        >
          <i
            class="fa-solid fa-up-right-from-square"
          />
        </a>
      </td>
    </tr>,
  },
]
`;
